Make it easy to develop with broadcast services

benthorner opened the pull request:
Previously I had to handcraft some SQL to give myself access to a
broadcast service I created locally. I've done this enough times
that I think it's worth automating.

klssmith commented:
It feels like this could be a bit dangerous for a couple of reasons. It will put your local development environment in a state which isn't possible "normally" and there's a chance that could lead to us making errors when writing or testing new code locally - users wouldn't ever be able to have all the permissions in [this list](https://github.com/alphagov/notifications-api/blob/b2b921ef21e9b88fe7018af60f95aca81d9abfd2/app/models.py#L1869).

My other concern is that there is nothing to stop this command being run in a non-dev environment. Even though there is alerting in place, we maybe need to check if it's considered too risky to add this since it would provide an quick and easy way to give someone a lot of permissions while bypassing some of the audit trails.

benthorner commented:
Thanks for taking a look @klssmith. I put this up as a quick attempt but as you point out it has several potential flaws. I've now spent some more time on it and have replaced the original commit and added some tests in another. I've also added an optional, final refactor commit to fix a surprise I found when writing the tests.

> It will put your local development environment in a state which isn't possible "normally"

Agreed. That could lead to a confusing discrepancy if the permission code had a bug in it. I've changed it to just populate the permissions for a broadcast creator + approver, as well as restricting the change to just broadcast services.

> My other concern is that there is nothing to stop this command being run in a non-dev environment.

I'm not concerned about this as if you're able to run this command in Production then you can also run any code you like. However, it's good practice to be defensive and document more clearly where code is meant to be used, so I've added an environment restriction to it so it can only be run locally (test + development).

Would you like to take another look?

klssmith commented on app/commands.py:
```diff
@@ -914,3 +925,32 @@ def populate_annual_billing_with_defaults(year, missing_services_only):
+
+    user_broadcast_services = Service.query.filter(
+        Service.permissions.any(permission='broadcast'),
+        Service.users.any(id=user_id)
```
You could consider another filter to ensure that only active services are changed

klssmith commented on tests/app/test_commands.py:
```diff
@@ -0,0 +1,24 @@
+    sample_service,
+    sample_broadcast_service,
+    notify_api,
+    mocker,
```
`mocker` is not used in this test

benthorner replied:
Good spot - I've amended the second commit to remove it.

klssmith commented on tests/app/conftest.py:
```diff
@@ -828,22 +825,21 @@ def letter_volumes_email_template(notify_service):
 @pytest.fixture
-def notify_service(notify_db, notify_db_session):
-    user = create_user()
+def notify_service(notify_db, notify_db_session, sample_user):
```
Only the `sample_user` fixture is needed for this test.

`sample_user` includes `notify_db_session` which in turn includes `notify_db`. Basically, we can remove some fixtures because none of the tests which have been updated to now use `sample_user` need `notify_db_session` or `notify_db`.

benthorner replied:
Nice (grumbles about dependency injection). I've added an extra commit to remove all the redundant DB fixtures.

klssmith reviewed (commented):
I left a couple of very minor comments, but this looks good - nice job getting testing the command file to work!
